File: lesson_12_20241105/binary_tree.js
class Node {
    constructor(value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }
}


class BinaryTree {
    constructor() {
        this.root = null;
    }   
    
    getRoot() {
        return this.root;
    }
   

    print(data) {
        console.log(data);
    }

    insert(value) {
        const newNode = new Node(value);
        if (this.root === null) { //пустое дерево
            this.root = newNode;
        } else {
            let current = this.root;
            let parent;
            
            while (true) {
                parent = current;
                if (value === current.value) { // такой объект уже есть
                    return;
                } else if(value < current.value) {
                    current = current.left;
                    if(current === null) {
                        parent.left = newNode;
                        return;
                    }
                } else { //value > current.value
                    current = current.right;
                    if(current === null) {
                        parent.right = newNode;
                        return;
                    }
                }
            }

        }
    }

     // вставка через рекурсию
    insertRecursively(value, node = this.root)
    {
        if (!node) {
            return new Node(value);
        }

        if (value < node.value)
         {
            node.left = this
            .insertRecursively(value, node.left);
        } else if (value > node.value) 
        {
            node.right = this
            .insertRecursively(value, node.right);
        }

        return node;
    }

    find(value) {
        let current = this.root;

        while (current.value !== value) {
            if (current.value > value) {
                current = current.left;
            } else {
                current = current.right;
            }
            if (current === null) {
                return null;
            }
        }
        return current;
    }
}

// Create a binary tree
const bt = new BinaryTree();

// 3,1,2,4
// bt.insert(3);
// bt.insert(1);
// bt.insert(2);
// bt.insert(4);
// bt.print(bt.root);

bt.root = bt.insertRecursively(3);
bt.insertRecursively(1);
bt.insertRecursively(2);
bt.insertRecursively(4);

bt.print(bt.root);

console.log(bt.find(2));
console.log(bt.find(5));

File: lesson_12_20241105/BinaryTree.java
package lesson_12_20241105;
// Реализуйте методы: 
// find(value)
// insert(value)
// delete(value)
// print(node)

public class BinaryTree {
    private Node root;

    public BinaryTree(Node root) {
        this.root = root;
    }

    public BinaryTree() {
    }

    public Node getRoot() {
        return root;
    }

    public void setRoot(Node root) {
        this.root = root;
    }

    public void print() {
        System.out.println(root);
    }

    public void insert(int value) {
        Node newNode = new Node();
        newNode.setValue(value);

        if(root == null) {
            root = newNode;
        } else {
            Node current = root;
            Node parent = null;

            while(true) {
                parent = current;
                if(value == current.getValue()) { // такой узел уже есть
                    return;
                } else if(value < current.getValue()) {
                    current = current.getLeft();
                    if(current == null) {
                        parent.setLeft(newNode);
                        return;
                    }
                } else {
                    current = current.getRight();
                    if(current == null) {
                        parent.setRight(newNode);
                        return;
                    }
                }
            }
        }
    }

   // вставка через рекурсию
   public Node insertRecursive(Node current, Node newNode) {
        if(root == null) {
            root = newNode;
            return newNode;
        }
        if(current == null) {
            return newNode;
        }

        if(newNode.getValue() < current.getValue()) {
            current.setLeft(insertRecursive(current.getLeft(), newNode));
        } else if(newNode.getValue() > current.getValue()) {
            current.setRight(insertRecursive(current.getRight(), newNode));
        } else {
            return current;
        }
        return current;
    }

    public Node find(int value) {
        Node current = root; // начинаем поиск с корневого узла

        while (current.getValue() != value) { // поиск пока не будет найден элемент или не будут перебраны все
            if (current.getValue() > value) { // движение влево?
                current = current.getLeft();
            } else {
                current = current.getRight(); //движение вправо
            }
            if (current == null) { // если потомка нет,
                return null;
            }
        }
        return current; // возвращаем найденный элемент
    }

   public boolean delete(int value) 
   {
       Node currentNode = root;
       Node parentNode = root;
       boolean isLeftChild = true;
       // ищем элемент в дереве по значению
       while (currentNode.getValue() != value) { 
           parentNode = currentNode;
           if (value < currentNode.getValue()) { 
               isLeftChild = true;
               currentNode = currentNode.getLeft();
           }
           else { 
               isLeftChild = false;
               currentNode = currentNode.getRight();
           }
           if (currentNode == null)
               return false; 
       }

       
       if (currentNode.getLeft() == null && currentNode.getRight() == null) {  // это лист
           if (currentNode == root) 
               root = null;
           else if (isLeftChild)
               parentNode.setLeft(null); 
           else
               parentNode.setRight(null);
       }
       else if (currentNode.getRight() == null) {  //справа нет дочерних
           if (currentNode == root)
               root = currentNode.getLeft();
           else if (isLeftChild)
               parentNode.setLeft(currentNode.getLeft());
           else
               parentNode.setRight(currentNode.getLeft());
       }
       else if (currentNode.getLeft() == null) {  //слева нет дочерних
           if (currentNode == root)
               root = currentNode.getRight();
           else if (isLeftChild)
               parentNode.setLeft(currentNode.getRight());
           else
               parentNode.setRight(currentNode.getRight());
       }
       else { // и справа и слева есть дочерние элементы
           Node heir = receiveHeir(currentNode);
           if (currentNode == root)
               root = heir;
           else if (isLeftChild)
               parentNode.setLeft(heir);
           else
               parentNode.setRight(heir);
       }
       return true; 
   }
   
   private Node receiveHeir(Node node) {
       Node parentNode = node;
       Node heirNode = node;
       Node currentNode = node.getRight(); 
       while (currentNode != null) 
       {
           parentNode = heirNode;
           heirNode = currentNode;
           currentNode = currentNode.getLeft(); 
       }
       
       if (heirNode != node.getRight()) 
       { 
           parentNode.setLeft(heirNode.getRight());
           heirNode.setRight(node.getRight());
       }
       return heirNode;
   }

}
